src/Featurizers/Components/NormUpdaters.h

```diff
@@ -0,0 +1,413 @@
+// ----------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License
+// ----------------------------------------------------------------------
+#pragma once
+
+#include <limits>
+#include "../../Traits.h"
+namespace Microsoft {
+namespace Featurizer {
+namespace Featurizers {
+namespace Components {
+
+namespace TypeSelector {
+// max norm type selector is created for selecting different types for max norm based on input type
+// when input type is integer, uint64 is chosen for performance
+// when input type is numeric, long double is chosen
+template <typename T>
+struct MaxNormTypeSelector {
+    using type = std::uint64_t;
+};
+template <>
+struct MaxNormTypeSelector<std::float_t> {
+    using type = long double;
+};
+
+template <>
+struct MaxNormTypeSelector<std::double_t> {
+    using type = long double;
+};
+
+}
+
+#if (defined __clang__)
+#   pragma clang diagnostic push
+#   pragma clang diagnostic ignored "-Wfloat-equal"
+#   pragma clang diagnostic ignored "-Wdouble-promotion"
+#endif
+
+namespace Details{
+
+/////////////////////////////////////////////////////////////////////////
+///  \class         L1NormUpdater
+///  \brief         L1NormUpdater updates l1 norm
+///                 l1 norm is the sum of elements absolute value
+///
+template <typename T>
+class L1NormUpdater {
+public:
+
+    /////////////////////////////////////////////////////////////////////////
+    ///  \class         L1NormResult
+    ///  \brief         A structure which contains the l1 norm
+    ///
+    struct L1NormResult {

```

Comment:
Nit: The struct name can be Result since it will always have to be referred to via its qualified name (`L1NormUpdater::L1NormResult` vs `LiNormUpdater::Result`).